test/unit/puppetlabs/services/jruby/jruby_pool_test.clj

```diff
@@ -99,6 +99,62 @@
               (doseq [k (keys counts)]
                 (is (= (inc (counts k)) (new-counts k)))))))))))
 
+(deftest borrow-while-pool-is-being-initialized-test
+  (testing "borrow will block until an instance is available while the pool is coming online"
+    (let [pool-initialized? (promise)
+          init-fn (fn [instance] @pool-initialized? instance)
+          pool-size 1
+          config (jruby-testutils/jruby-config
+                  {:max-active-instances pool-size
+                   :lifecycle {:initialize-pool-instance init-fn}})
+          pool-context (jruby-core/create-pool-context config)]
+
+      ;; start a pool initialization, which will block on the `initialize-pool-instance`
+      ;; function's deref of the promise
+      (jruby-agents/send-prime-pool! pool-context)
```

Comment:
curious why you aren't using a service here and calling `create-pool`?  As-is it seems like this test relies on knowledge of what happens inside that function?

Also this makes me wonder if `create-pool-context` should be in `core` or somewhere else?  Do we expect users to call it?